## Lending Club Credit Risk Modeling
- In this project, I will build three **machine learning** models to predict the three components of expected loss in the context of **credit risk modeling** at the **Lending Club** (a peer-to-peer credit company): Probability of Default (PD), Exposure at Default (EAD) and Loss Given Default (LGD). The expected loss will be the product of these elements: **Expected Loss (EL) = PD * EAD * LGD.** The PD Model will be used to stablish a credit policy, deciding wheter to grant a loan or not for new applicants (application model) based on their credit scores. The EAD and LGD Models will be used along with the PD Model to estimate the Expected Loss (EL) from each loan in order to assess the capital the institution needs to hold to protect itself against defaults.
- The PD modelling encompasses a binary classification problem with target being 1 in case of non-default and 0 in case of default. A Logistic Regression model will be built. 
- The LGD and EAD modelling encompasses a beta regression problem, that is, a regression task in which the dependent variable is beta distributed.
- The **solution pipeline** is based on the **crisp-dm** framework:
    1. Business understanding.
    2. Data understanding.
    3. Data preparation.
    4. Modelling.
    5. Validation.
    6. Deployment.
- Below, I will introduce **information** about the **company**, the **business problem**, the **project objectives and benefits**, and some important **concepts** to have in mind, given that I am dealing with a credit risk problem.

## Lending Club
- **What is the Lending Club?:**
    - LendingClub is a **peer-to-peer lending platform** that facilitates the borrowing and lending of money directly between individuals, without the need for traditional financial institutions such as banks. The platform operates as an online marketplace, connecting borrowers seeking personal loans with investors willing to fund those loans.

## Business Problem: Managing Default Risks and Optimizing Returns
- **What is the business problem?:**
    - LendingClub faces a significant business challenge related to **managing default risks effectively** while **optimizing returns** for its investors. The platform facilitates peer-to-peer lending, connecting borrowers with investors, and relies on **accurate risk assessments to maintain a sustainable and profitable lending ecosystem.** Thus, the CEO wants us to provide insights about which factors are associated with credit risk in Lending Club's operations, and to construct models capable of predicting the probability of default for new applicants and possible losses on its loans in order to establish a credit policy, deciding when to grant a loan or not for an applicant.
- **Which are the project objectives and benefits?**
    1. Identify the factors associated with credit risk in the form of business insights.
    2. Develop an accurate Probability of Default (PD) Model, constructing a scorecard. This will allow Lending Club to decide wheter to grant a loan or not to a new applicant (application model), based on credit scores.
    3. Develop Exposure at Default (EAD) and Loss Given Default (LGD) Models, to estimate the Expected Loss in loans. This will allow Lending Club to hold sufficient capital to protect itself against default in each loan.
    4. Improve risk management and optimize returns by selecting a cut-off point or creating risk categories to approve loans based on our models' results. With this, we intend to establish a credit policy, trying to balance risk and ROI of Lending Club's assets.
    5. Apply model monitoring and maintenance techniques to safeguard our results from population instability, characterized by significant changes in loan applicants' characteristics. This will allow us to understand whether the built model is still useful in the future or whether the loan applicants characteristics changed significantly, such that we will need to redevelop it.

## Credit Risk Modeling Concepts: The Context of the Problem
- **Financial institutions**, like LendingClub and online lending platforms, **make money by lending to people and businesses.** When they lend money, they **charge interest**, which is a significant source of their **profits**. **Managing credit risk well is crucial** for these institutions. This means ensuring that borrowers pay back their loans on time to avoid losses.
- **Credit risk** is the possibility that a borrower might not fulfill their financial obligations, leading to a loss for the lender. If a borrower fails to meet the agreed-upon terms, it's called a "default," and it can result in financial losses for the lender. The **default** definition is associated with a time horizon. For example, if a borrower hasn't paid their debt within 90 days of the due date, they are considered in default.
- In the credit market, important **rules** help keep things honest and clear. **Basel III** is one such set of rules, making sure banks have **enough money (capital requirements)** and follow **guidelines for assessing loan risks**. The **Internal Rating-Based Approach (IRB-A)** lets banks figure out credit risks using concepts like Probability of Default (PD), Exposure at Default (EAD), and Loss Given Default (LGD). Another rule, **International Financial Reporting Standard 9 (IFRS 9)**, gives standards for measuring financial assets. It's special because it looks at the chance of a loan not being paid back over its entire life, unlike Basel, which checks it for one year. These rules help banks have enough money, handle risks well, and keep the credit market steady and trustworthy.
- The **"expected loss (EL)"** is the average estimated loss that a lender can expect from loans that default. It involves three factors: the **probability of default (likelihood of a borrower defaulting)**, **loss given default (portion of the amount the bank is exposed to that can't be recovered in case of default)**, and **exposure at default (potential loss at the time of default, considering the outstanding loan amount and other factors)**.
- **LendingClub**, operating as a peer-to-peer lending platform, uses a **"PD Model/Credit Scoring Model" to assess borrowers' creditworthiness using credit scores** from a scorecard. This helps determine the **likelihood of loan repayment**, guiding the decision to **approve or deny the loan.** The **required capital to guard against default** for each loan is calculated using **EAD and LGD Models** to estimate the **Expected Loss (EL)**, contributing to minimizing risk in credit operations.
- When creating a Credit Scoring Model, which assesses creditworthiness for loan approval, using data available at the time of the application is considered an **"application model."** It is distinct from a "behavior model." This is the model I will build here.
- A **"credit policy"** is a set of guidelines that financial institutions follow to evaluate and manage lending risk. It involves setting a cut-off point, a threshold based on credit scores, or risk categories, to decide whether to approve or reject loan applications. Another factors such as the expected ROI for each loan application are included.
- **"Return on Investment (ROI)"** is a key measure of loan profitability. Balancing ROI with risk is vital for effective credit policy management. While higher-risk loans may offer more significant potential returns, they also come with a higher chance of default.

### 0. Importing the libraries

In [1]:
# Data manipulation and visualization.
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns

# Display options.
pd.set_option('display.max_rows', None)
pd.set_option('display.max_info_rows', 100)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)

### 1. Data Understanding
- The data was collected from kaggle and contain complete loan data for all loans issued through the 2007-2015, including the current loan status (Current, Late, Fully Paid, etc.) and the latest payment information.
- This include consumer loans, as explained above in Lending Club's services, and encompasses data available at the moment of the application. Thus, it will be used to build an application model.
- In this step, I intend to perform some initial data understanding, like assessing the dimensions of the dataset, features data types, missing and duplicated values, categorical and numerical variables and so on.

In [2]:
path = '/Users/pedrohenriquealmeidaoliveira/Documents/data_science/large_files/loan_data_2007_2014.csv'
df = pd.read_csv(path)

/var/folders/7m/gml801pj3c12rljtszckgvr00000gn/T/ipykernel_8227/3754117770.py:2: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(path)


In [3]:
df.head()

Unnamed: 0       id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment grade sub_grade                 emp_title emp_length home_ownership  annual_inc verification_status issue_d  loan_status pymnt_plan                                                url                                               desc         purpose                 title zip_code addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  total_acc initial_list_status  out_prncp  out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d last_credit_pull_d  collections_12_mths_ex_med  mths_since_last_major_derog  policy_code application_type  annual_inc_joint  dti_joint  verification_status_joint  acc_now_delinq  tot_coll_amt  tot_cur_bal  open_acc_6m  open_il_6m   
0           0  1077501    1296599       5000         5000           4975.0   36 months     10.65       162.87     B        B2                       NaN  10+ years           RENT     24000.0            Verified  Dec-11   Fully Paid          n  https://www.lendingclub.com/browse/loanDetail....    Borrower added on 12/22/11 > I need to upgra...     credit_card              Computer    860xx         AZ  27.65          0.0           Jan-85             1.0                     NaN                     NaN       3.0      0.0      13648        83.7        9.0                   f        0.0            0.0   5861.071414          5831.78          5000.00         861.07                0.00        0.00                     0.00       Jan-15           171.62          NaN             Jan-16                         0.0                          NaN            1       INDIVIDUAL               NaN        NaN                        NaN             0.0           NaN          NaN          NaN         NaN  \
1           1  1077430    1314167       2500         2500           2500.0   60 months     15.27        59.83     C        C4                     Ryder   < 1 year           RENT     30000.0     Source Verified  Dec-11  Charged Off          n  https://www.lendingclub.com/browse/loanDetail....    Borrower added on 12/22/11 > I plan to use t...             car                  bike    309xx         GA   1.00          0.0           Apr-99             5.0                     NaN                     NaN       3.0      0.0       1687         9.4        4.0                   f        0.0            0.0   1008.710000          1008.71           456.46         435.17                0.00      117.08                     1.11       Apr-13           119.66          NaN             Sep-13                         0.0                          NaN            1       INDIVIDUAL               NaN        NaN                        NaN             0.0           NaN          NaN          NaN         NaN   
2           2  1077175    1313524       2400         2400           2400.0   36 months     15.96        84.33     C        C5                       NaN  10+ years           RENT     12252.0        Not Verified  Dec-11   Fully Paid          n  https://www.lendingclub.com/browse/loanDetail....                                                NaN  small_business  real estate business    606xx         IL   8.72          0.0           Nov-01             2.0                     NaN                     NaN       2.0      0.0       2956        98.5       10.0                   f        0.0            0.0   3003.653644          3003.65          2400.00         603.65                0.00        0.00                     0.00       Jun-14           649.91          NaN             Jan-16                         0.0                          NaN            1       INDIVIDUAL               NaN        NaN                        NaN             0.0           NaN          NaN          NaN         NaN   
3          

In [4]:
df.tail()

Unnamed: 0       id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  installment grade sub_grade                              emp_title emp_length home_ownership  annual_inc verification_status issue_d  loan_status pymnt_plan                                                url                                               desc             purpose                    title zip_code addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  total_acc initial_list_status  out_prncp  out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt next_pymnt_d last_credit_pull_d  collections_12_mths_ex_med  mths_since_last_major_derog  policy_code application_type  annual_inc_joint  dti_joint  verification_status_joint  acc_now_delinq  tot_coll_amt  tot_cur_bal   
466280      466280  8598660    1440975      18400        18400          18400.0   60 months     14.47       432.64     C        C2                     Financial Advisor     4 years       MORTGAGE    110000.0     Source Verified  Jan-14      Current          n  https://www.lendingclub.com/browse/loanDetail....                                                NaN  debt_consolidation       Debt consolidation    773xx         TX  19.85          0.0           Apr-03             2.0                     NaN                     NaN      18.0      0.0      23208        77.6       36.0                   w   12574.00       12574.00  10383.360000         10383.36          5826.00        4557.36                 0.0         0.0                      0.0       Jan-16           432.64       Feb-16             Jan-16                         0.0                          NaN            1       INDIVIDUAL               NaN        NaN                        NaN             0.0           0.0     294998.0  \
466281      466281  9684700   11536848      22000        22000          22000.0   60 months     19.97       582.50     D        D5  Chief of Interpretation (Park Ranger)  10+ years       MORTGAGE     78000.0            Verified  Jan-14  Charged Off          n  https://www.lendingclub.com/browse/loanDetail....                                                NaN  debt_consolidation       Debt consolidation    377xx         TN  18.45          0.0           Jun-97             5.0                     NaN                   116.0      18.0      1.0      18238        46.3       30.0                   f       0.00           0.00   4677.920000          4677.92          1837.04        2840.88                 0.0         0.0                      0.0       Dec-14            17.50          NaN             Jan-16                         0.0                          NaN            1       INDIVIDUAL               NaN        NaN                        NaN             0.0           0.0     221830.0   
466282      466282  9584776   11436914      20700        20700          20700.0   60 months     16.99       514.34     D        D1                                 patrol    7 years       MORTGAGE     46000.0            Verified  Jan-14      Current          n  https://www.lendingclub.com/browse/loanDetail....    Borrower added on 12/06/13 > I am going to c...  debt_consolidation       Debt consolidation    458xx         OH  25.65          0.0           Dec-01             2.0                    65.0                     NaN      18.0      0.0       6688        51.1       43.0                   f   14428.31       14428.31  12343.980000         12343.98          6271.69        6072.29                 0.0         0.0                      0.0       Jan-16           514.34       Feb-16             Dec-15                         0.0                          NaN            1       INDIVIDUAL               NaN        NaN                        NaN             0.0           0.0      73598.0   
466283      4662

#### 1.1 Data dictionary
- Below we have the information about what each feature of the dataset means.

**addr_state**: The state provided by the borrower in the loan application

**annual_inc**: The self-reported annual income provided by the borrower during registration.

**annual_inc_joint**: The combined self-reported annual income provided by the co-borrowers during registration.

**application_type**: Indicates whether the loan is an individual application or a joint application with two co-borrowers

**collection_recovery_fee**: Post charge-off collection fee

**collections_12_mths_ex_med**: Number of collections in 12 months excluding medical collections

**delinq_2yrs**: The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years

**desc**: Loan description provided by the borrower

**dti**: A ratio calculated using the borrower’s total monthly debt payments on the total debt obligations, excluding mortgage and the requested LC loan, divided by the borrower’s self-reported monthly income.

**dti_joint**: A ratio calculated using the co-borrowers' total monthly payments on the total debt obligations, excluding mortgages and the requested LC loan, divided by the co-borrowers' combined self-reported monthly income

**earliest_cr_line**: The month the borrower's earliest reported credit line was opened

**emp_length**: Employment length in years. Possible values are between 0 and 10 where 0 means less than one year and 10 means ten or more years.

**emp_title**: The job title supplied by the Borrower when applying for the loan.

**fico_range_high**: The upper boundary range the borrower’s FICO at loan origination belongs to.

**fico_range_low**: The lower boundary range the borrower’s FICO at loan origination belongs to.

**funded_amnt**: The total amount committed to that loan at that point in time.

**funded_amnt_inv**: The total amount committed by investors for that loan at that point in time.

**grade**: LC assigned loan grade

**home_ownership**: The home ownership status provided by the borrower during registration. Our values are: RENT, OWN, MORTGAGE, OTHER.

**id**: A unique LC assigned ID for the loan listing.

**initial_list_status**: The initial listing status of the loan. Possible values are – W, F

**inq_last_6mths**: The number of inquiries in the past 6 months (excluding auto and mortgage inquiries)

**installment**: The monthly payment owed by the borrower if the loan originates.

**int_rate**: Interest Rate on the loan

**is_inc_v**: Indicates if income was verified by LC, not verified, or if the income source was verified

**issue_d**: The month which the loan was funded

**last_credit_pull_d**: The most recent month LC pulled credit for this loan

**last_fico_range_high**: The upper boundary range the borrower’s last FICO pulled belongs to.

**last_fico_range_low**: The lower boundary range the borrower’s last FICO pulled belongs to.

**last_pymnt_amnt**: Last total payment amount received

**last_pymnt_d**: Last month payment was received

**loan_amnt**: The listed amount of the loan applied for by the borrower. If at some point in time, the credit department reduces the loan amount, then it will be reflected in this value.

**loan_status**: Current status of the loan

**member_id**: A unique LC assigned Id for the borrower member.

**mths_since_last_delinq**: The number of months since the borrower's last delinquency.

**mths_since_last_major_derog**: Months since the most recent 90-day or worse rating

**mths_since_last_record**: The number of months since the last public record.

**next_pymnt_d**: Next scheduled payment date

**open_acc**: The number of open credit lines in the borrower's credit file.

**out_prncp**: Remaining outstanding principal for the total amount funded

**out_prncp_inv**: Remaining outstanding principal for the portion of total amount funded by investors

**policy_code**: Publicly available policy_code=1; new products not publicly available policy_code=2

**pub_rec**: Number of derogatory public records

**purpose**: A category provided by the borrower for the loan request.

**pymnt_plan**: Indicates if a payment plan has been put in place for the loan

**recoveries**: Post charge-off gross recovery

**revol_bal**: Total credit revolving balance

**revol_util**: Revolving line utilization rate, or the amount of credit the borrower is using relative to all available revolving credit.

**sub_grade**: LC assigned loan subgrade

**term**: The number of payments on the loan. Values are in months and can be either 36 or 60.

**title**: The loan title provided by the borrower

**total_acc**: The total number of credit lines currently in the borrower's credit file

**total_pymnt**: Payments received to date for the total amount funded

**total_pymnt_inv**: Payments received to date for the portion of the total amount funded by investors

**total_rec_int**: Interest received to date

**total_rec_late_fee**: Late fees received to date

**total_rec_prncp**: Principal received to date

**url**: URL for the LC page with listing data

**verified_status_joint**: Indicates if the co-borrowers' joint income was verified by LC, not verified, or if the income source was verified

**zip_code**: The first 3 numbers of the zip code provided by the borrower in the loan application.

**open_acc_6m**: Number of open trades in the last 6 months

**open_il_6m**: Number of currently active installment trades

**open_il_12m**: Number of installment accounts opened in the past 12 months

**open_il_24m**: Number of installment accounts opened in the past 24 months

**mths_since_rcnt_il**: Months since most recent installment accounts opened

**total_bal_il**: Total current balance of all installment accounts

**il_util**: Ratio of total current balance to high credit/credit limit on all installment accounts

**open_rv_12m**: Number of revolving trades opened in the past 12 months

**open_rv_24m**: Number of revolving trades opened in the past 24 months

**max_bal_bc**: Maximum current balance owed on all revolving accounts

**all_util**: Balance to credit limit on all trades

**total_rev_hi_lim**: Total revolving high credit/credit limit

**inq_fi**: Number of personal finance inquiries

**total_cu_tl**: Number of finance trades

**inq_last_12m**: Number of credit inquiries in the past 12 months

**acc_now_delinq**: The number of accounts on which the borrower is now delinquent.

**tot_coll_amt**: Total collection amounts ever owed

**tot_cur_bal**: Total current balance of all accounts

*Employer Title replaces Employer Name for all loans listed after 9/23/2013*

- From the data dictionary above, we can see that our **dependent variable for the PD model**, a flag indicating 1 in case of non-default and 0 in case of default will be obtained from the **loan_status** variable.
- At a glance, we can divide our variables into similar **context/information groups:**
    - **Demographical Variables:** addr_state, annual_inc, application_type, emp_length, emp_title, home_ownership, title, zip_code
    - **Financial Variables:** annual_inc_joint, collection_recovery_fee, dti, dti_joint, funded_amnt, funded_amnt_inv, installment, int_rate, loan_amnt, revol_bal, revol_util, total_acc
    - **Credit and Risk Variables:** delinq_2yrs, fico_range_high, fico_range_low, grade, initial_list_status, inq_last_6mths, inq_last_12m, last_fico_range_high, last_fico_range_low, loan_status, pub_rec, sub_grade, term
    - **Time-Related and Other Variables:** desc, issue_d, earliest_cr_line, id, last_credit_pull_d, last_pymnt_amnt, last_pymnt_d, mths_since_last_delinq, mths_since_last_major_derog, mths_since_last_record, next_pymnt_d, policy_code, purpose, pymnt_plan, recoveries, total_pymnt, total_pymnt_inv, total_rec_int, total_rec_late_fee, total_rec_prncp, url, verified_status_joint, open_acc_6m, open_il_6m, open_il_12m, open_il_24m, mths_since_rcnt_il, total_bal_il, il_util, open_rv_12m, open_rv_24m, max_bal_bc, all_util, total_rev_hi_lim, inq_fi, total_cu_tl, acc_now_delinq, tot_coll_amt, tot_cur_bal 
- At a glance, there are some **variables** that can drive **valuable insights** about **credit risk** in Lending Club's operations:
    - Most of the demographical variables: addr_state, annual_inc, emp_length, home_ownership.
    - From the financial variables: dti, funded_amnt, installment, int_rate, loan_amnt, revol_bal, total_acc.
    - From the credit and risk variables: delinq_2yrs, fico scores variables, inq_last_6_months and 12m, and term.
    - From time-related and other variables: issue_d, purpose, tot_cur_bal, total_rev_hi_lim.
- At a glance, some variables just may not be useful and can be removed. It is the case for url, for example, but I will analyze this further in the notebook. Moreover, some variables like loan_amnt will not be available at the moment of prediction (in the production environment) and thus will be also dropped in order to prevent data leakage. Considering these features can provide valuable insights about credit risk in Lending Club's operations, they will be removed just in the data preprocessing step, once we have already explored them. 

#### 1.2 General information
- I will assess general information about the data, such as dimensions, features data types, null and duplicated values, and so on in order to initially see some possible problems that will require data cleaning and gather more information about the data I am dealing with.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 466285 entries, 0 to 466284
Data columns (total 75 columns):
 #   Column                       Dtype  
---  ------                       -----  
 0   Unnamed: 0                   int64  
 1   id                           int64  
 2   member_id                    int64  
 3   loan_amnt                    int64  
 4   funded_amnt                  int64  
 5   funded_amnt_inv              float64
 6   term                         object 
 7   int_rate                     float64
 8   installment                  float64
 9   grade                        object 
 10  sub_grade                    object 
 11  emp_title                    object 
 12  emp_length                   object 
 13  home_ownership               object 
 14  annual_inc                   float64
 15  verification_status          object 
 16  issue_d                      object 
 17  loan_status                  object 
 18  pymnt_plan                   object 
 19  ur

In [6]:
print(f'The dataset has {df.shape[0]} rows and {df.shape[1]} columns.')

The dataset has 466285 rows and 75 columns.


- We have 46 float variables, 7 int variables and 22 object variables, totalizing 53 numerical features and 22 categorical features.
- At a glance, **some variables can be removed.** This include id variables (because they have one unique value per observation), url and desc (because they aren't useful). 
- **Some variables are in the incorrect data type** and must be converted to the correct one. It is the case of term and emp_length, which must be converted from object to int. Moreover, issue_d, earliest_cr_line, last_pymnt_d, next_pymnt_d and last_credit_pull_d must be converted from object to datetime format, because they are date-related features.

#### 1.3 Missing and Duplicate values
- I will assess the missing values distributions across the variables in order to investigate their motivation and treat them further. The same for duplicate values.

In [3]:
missing_df = df.isna().sum().to_frame().rename(columns={0: 'count'}).sort_values(by=['count'], ascending=False)
missing_df['pct'] = round(missing_df['count'] / df.shape[0] * 100, 3)
missing_df

count      pct
inq_last_12m                 466285  100.000
total_bal_il                 466285  100.000
dti_joint                    466285  100.000
verification_status_joint    466285  100.000
annual_inc_joint             466285  100.000
open_acc_6m                  466285  100.000
open_il_6m                   466285  100.000
open_il_12m                  466285  100.000
open_il_24m                  466285  100.000
mths_since_rcnt_il           466285  100.000
il_util                      466285  100.000
open_rv_24m                  466285  100.000
total_cu_tl                  466285  100.000
inq_fi                       466285  100.000
max_bal_bc                   466285  100.000
all_util                     466285  100.000
open_rv_12m                  466285  100.000
mths_since_last_record       403647   86.567
mths_since_last_major_derog  367311   78.774
desc                         340304   72.982
mths_since_last_delinq       250351   53.691
next_pymnt_d                 227214   48.729
tot_cur_bal                   70276   15.071
tot_coll_amt                  70276   15.071
total_rev_hi_lim              70276   15.071
emp_title                     27588    5.917
emp_length                    21008    4.505
last_pymnt_d                    376    0.081
revol_util                      340    0.073
collections_12_mths_ex_med      145    0.031
last_credit_pull_d               42    0.009
inq_last_6mths                   29    0.006
earliest_cr_line                 29    0.006
delinq_2yrs                      29    0.006
open_acc                         29    0.006
pub_rec                          29    0.006
acc_now_delinq                   29    0.006
total_acc                        29    0.006
title                            21    0.005
annual_inc                        4    0.001
application_type                  0    0.000
policy_code                       0    0.000
collection_recovery_fee           0    0.000
last_pymnt_amnt                   0    0.000
Unnamed: 0                        0    0.000
recoveries                        0    0.000
total_rec_late_fee                0    0.000
member_id                         0    0.000
loan_amnt                         0    0.000
funded_amnt                       0    0.000
funded_amnt_inv                   0    0.000
term                              0    0.000
int_rate                          0    0.000
installment                       0    0.000
grade                             0    0.000
sub_grade                         0    0.000
home_ownership                    0    0.000
verification_status               0    0.000
issue_d                           0    0.000
loan_status                       0    0.000
pymnt_plan                        0    0.000
url                               0    0.000
purpose                           0    0.000
zip_code                          0    0.000
addr_state                        0    0.000
dti                               0    0.000
revol_bal                         0    0.000
initial_list_status               0    0.000
id                                0    0.000
out_prncp_inv                     0    0.000
total_pymnt                       0    0.000
total_pymnt_inv                   0    0.000
total_rec_prncp                   0    0.000
total_rec_int                     0    0.000
out_prncp                         0    0.000

In [8]:
df.duplicated().sum()

0

- There are not duplicated values in the data.
- It is possible to see that there are a lot of variables with a 100% of missings. These are not useful for anything and will be removed.
- I will stablish a 70% cut-off to remove variables with missing values. That is, if the variable has 70% or more missing rate, it will be removed.
- For the other features, I will investigate the missing values occurrence and decide whether to keep them or not. Some considerations:
    - If the **pattern of missing values is not random**, meaning that the occurrence of missing values varies between good and bad loans, for example, if there are proportionally more missing values in bad loans, it is advisable to treat it as a category within the feature. The absence of information from the applicant could be related to their intention to repay or not repay their debt.
    - In case the **pattern of missing values is random, we can:**
        - Remove these observations, in case we have a big sample of applicant data (and we have!) and their frequency is really low.
        - Exclude variables with an excessive number of missing values (which we already considered doing).
        - Keep the missing as one of the categories of the variable, like "unknown". It will have a weight in the final scoring formula.
        - Impute values. Utilizing mean, median, mode or even inferring it from another variable highly linearly correlated with the variable with missings are possible alternatives.
- The occurrence of **missing values** might be due to different **reasons**, like:
    - The applicant didn't gave this information.
    - The given information was not registered.
    - The given information does not match with any of the expected codes or values for that variable, then, it was registered as a missing. This is pretty common in occupation, where some abbreviations come.
    - The missing is, in reality, zero.
- All the referred treatments and investigations will be performed in the data preprocessing step.

#### 1.4 Numerical and Categorical features
- I'll check which numerical and categorical variables are in the data. This helps me separate them and check how many different categories there are in the categorical features. The number of categories is important because it affects how we encode the data. If there are too many categories, using one-hot encoding can make the data dimensionality too big. This can be a problem. Also, I'll see if a variable is helpful or not. Sometimes, variables have too many categories and don't really help the model. In that case, we might need to remove them from the analysis.

In [4]:
# Assessing which are the numerical and categorical features.
numerical_features = df.select_dtypes('number').columns.tolist()
categorical_features = df.select_dtypes('object').columns.tolist()
print(f'There are {len(numerical_features)} in the data. They are: {numerical_features}.')
print(f'There are {len(categorical_features)} in the data. They are: {categorical_features}.')

There are 53 in the data. They are: ['Unnamed: 0', 'id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'int_rate', 'installment', 'annual_inc', 'dti', 'delinq_2yrs', 'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_amnt', 'collections_12_mths_ex_med', 'mths_since_last_major_derog', 'policy_code', 'annual_inc_joint', 'dti_joint', 'verification_status_joint', 'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m', 'open_il_6m', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc', 'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m'].
There are 22 in the data. They are: ['term', 'grade', 'sub_grade', 'emp_title', 'emp_length'

In [13]:
# Assessing categorical features cardinality.
for feature in categorical_features:
    print(f'{feature} - cardinality = {df[feature].nunique()}')
    print()
    print(df[feature].unique())
    print('-' * 100)

term - cardinality = 2

[' 36 months' ' 60 months']
----------------------------------------------------------------------------------------------------
grade - cardinality = 7

['B' 'C' 'A' 'E' 'F' 'D' 'G']
----------------------------------------------------------------------------------------------------
sub_grade - cardinality = 35

['B2' 'C4' 'C5' 'C1' 'B5' 'A4' 'E1' 'F2' 'C3' 'B1' 'D1' 'A1' 'B3' 'B4'
 'C2' 'D2' 'A3' 'A5' 'D5' 'A2' 'E4' 'D3' 'D4' 'F3' 'E3' 'F4' 'F1' 'E5'
 'G4' 'E2' 'G3' 'G2' 'G1' 'F5' 'G5']
----------------------------------------------------------------------------------------------------
emp_title - cardinality = 205475

[nan 'Ryder' 'AIR RESOURCES BOARD' ... 'Mecánica'
 'Chief of Interpretation (Park Ranger)' 'Server Engineer Lead']
----------------------------------------------------------------------------------------------------
emp_length - cardinality = 11

['10+ years' '< 1 year' '1 year' '3 years' '8 years' '9 years' '4 years'
 '5 years' '6 years' '2 ye

In [5]:
# Assessing pymnt_plan and initial_list_status distributions in order to see whether there is imbalancement
df['pymnt_plan'].value_counts(normalize=True)

pymnt_plan
n    0.999981
y    0.000019
Name: proportion, dtype: float64

In [6]:
df['initial_list_status'].value_counts(normalize=True)

initial_list_status
f    0.649828
w    0.350172
Name: proportion, dtype: float64

#### 1.5 Data Cleaning and Feature Engineering
- I will apply an initial **data cleaning**, including the following tasks:
    - **Remove irrelevant variables:**
        - ids, url, desc, next_pymnt_d, application_type and pymnt_plan (almost constant, no variance), zip_code and title (high_cardinality) because they are not useful for neither the model or the analysis.
        - Variables with a missing rate higher than 70%.
        - As mentioned above, variables like loan_amnt will not be available at the moment of prediction (production env) and thus will be removed to prevent from data leakage. However, once they can present valuable insights about credit risk in Lending Club's operations, I will do this in the data preprocessing step, after exploring them.
    - **Convert variables to the correct data types:**
        - Convert term and emp_length from object to int data type.
        - Convert date realted features from object to datetime data type. Considering that I am using as reference data from 2007 to 2015, I will use 2015-01-01 as the current date for this feature extraction.
- From the features I have, I will try to **extract new features**, obtaining the maximum amount of information from the available variables.
    - credit_util_rate: Indicates the credit utilization rate, that is, how much of the available revolving credit the consumer is currently using. 
    - mths_since_earliest_cr_line: Self explanatory. It is interesting for us to obtain a numerical variable indicating the number of months since the earliest credit line and other date variables.

In [7]:
# Creating a copy to saveguard the original data.
clean_df = df.copy()

In [8]:
# Removing irrelevant variables.
to_drop_missing = missing_df.loc[missing_df['pct'] >= 70].index.tolist()
to_drop_irrelevant = ['id', 'member_id', 'url', 'Unnamed: 0', 'next_pymnt_d', 'application_type', 'pymnt_plan', 'zip_code', 'title']
to_drop = to_drop_missing + to_drop_irrelevant
clean_df = clean_df.drop(columns=to_drop)
clean_df.columns

Index(['loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'issue_d', 'loan_status', 'purpose', 'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths', 'mths_since_last_delinq', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'last_credit_pull_d', 'collections_12_mths_ex_med', 'policy_code', 'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'total_rev_hi_lim'], dtype='object')

In [11]:
# Converting variables to the correct data type.
clean_df['term'].unique()

array([' 36 months', ' 60 months'], dtype=object)

In [9]:
# Converting term variable to int format.
clean_df['term'] = clean_df['term'].apply(lambda x: int(x[1:3]))
clean_df['term'].unique(), clean_df.term.dtype

(array([36, 60]), dtype('int64'))

In [13]:
clean_df['emp_length'].unique()

array(['10+ years', '< 1 year', '1 year', '3 years', '8 years', '9 years',
       '4 years', '5 years', '6 years', '2 years', '7 years', nan],
      dtype=object)

In [10]:
# Converting emp_length to float format because of the missing values.
clean_df['emp_length'] = clean_df['emp_length'].str.replace(' years', '')
clean_df['emp_length'] = clean_df['emp_length'].str.replace('< 1 year', '0')
clean_df['emp_length'] = clean_df['emp_length'].str.replace(' year', '')
clean_df['emp_length'] = clean_df['emp_length'].str.replace('< ', '')
clean_df['emp_length'] = clean_df['emp_length'].str.replace('+', '')
clean_df['emp_length'] = clean_df['emp_length'].astype(float)
clean_df['emp_length'].unique()

array([10.,  0.,  1.,  3.,  8.,  9.,  4.,  5.,  6.,  2.,  7., nan])

In [15]:
clean_df['earliest_cr_line'].head()

0    Jan-85
1    Apr-99
2    Nov-01
3    Feb-96
4    Jan-96
Name: earliest_cr_line, dtype: object

In [11]:
# Converting earlist_cr_line to datetime format.
clean_df['earliest_cr_line'] = pd.to_datetime(clean_df['earliest_cr_line'], format='%b-%y')
clean_df['earliest_cr_line'].head()

0   1985-01-01
1   1999-04-01
2   2001-11-01
3   1996-02-01
4   1996-01-01
Name: earliest_cr_line, dtype: datetime64[ns]

In [17]:
clean_df['issue_d'].head()

0    Dec-11
1    Dec-11
2    Dec-11
3    Dec-11
4    Dec-11
Name: issue_d, dtype: object

In [12]:
# Converting issue_d to datetime format.
clean_df['issue_d'] = pd.to_datetime(clean_df['issue_d'], format='%b-%y')
clean_df['issue_d'].head()

0   2011-12-01
1   2011-12-01
2   2011-12-01
3   2011-12-01
4   2011-12-01
Name: issue_d, dtype: datetime64[ns]

In [16]:
#clean_df['last_pymnt_d'].head() 
clean_df['last_credit_pull_d'].head()

0    Jan-16
1    Sep-13
2    Jan-16
3    Jan-15
4    Jan-16
Name: last_credit_pull_d, dtype: object

In [13]:
# Extracting new features
CURR_DATE = pd.to_datetime('2015-01-01')

# Extracting a variable indicating the credit utilization rate. 
clean_df['credit_util_rate'] = (clean_df['revol_bal'] / clean_df['total_rev_hi_lim']) * 100

# Extracting mths_since_earliest_cr_line variable.
clean_df['mths_since_earliest_cr_line'] = round((CURR_DATE - clean_df['earliest_cr_line']) / np.timedelta64(1, 'M'))

# Extracting mths_since_issue_d variable.
clean_df['mths_since_issue_d'] = round((CURR_DATE - clean_df['issue_d']) / np.timedelta64(1, 'M'))

# Extracting mts_since_last_pymnt_d and mths_since_last_credit_pull_d variables.
clean_df['last_pymnt_d'] = pd.to_datetime(clean_df['last_pymnt_d'], format='%b-%y')
clean_df['last_credit_pull_d'] = pd.to_datetime(clean_df['last_credit_pull_d'], format='%b-%y')
clean_df['mths_since_last_pymnt_d'] = round((CURR_DATE - clean_df['last_pymnt_d']) / np.timedelta64(1, 'M'))
clean_df['mths_since_last_credit_pull_d'] = round((CURR_DATE - clean_df['last_credit_pull_d']) / np.timedelta64(1, 'M'))

In [14]:
# Dropping date variables after useful variables were extracted from them.
to_drop = ['earliest_cr_line', 'issue_d', 'last_pymnt_d', 'last_credit_pull_d']
clean_df = clean_df.drop(columns=to_drop)

#### 1.6 Descriptive Statistics
- I will assess some descriptive statistics of the variables in order to initially observe inconsistent information (when looking at minimum and maximum values), diagnose possible outliers and obtain initial simpler insights about the features distributions.

In [16]:
clean_df.describe().T

count           mean            std      min           25%           50%           75%           max
loan_amnt                      466285.0   14317.277577    8286.509164   500.00   8000.000000  12000.000000   20000.00000  3.500000e+04
funded_amnt                    466285.0   14291.801044    8274.371300   500.00   8000.000000  12000.000000   20000.00000  3.500000e+04
funded_amnt_inv                466285.0   14222.329888    8297.637788     0.00   8000.000000  12000.000000   19950.00000  3.500000e+04
term                           466285.0      42.605334      10.719040    36.00     36.000000     36.000000      60.00000  6.000000e+01
int_rate                       466285.0      13.829236       4.357587     5.42     10.990000     13.660000      16.49000  2.606000e+01
installment                    466285.0     432.061201     243.485550    15.67    256.690000    379.890000     566.58000  1.409990e+03
emp_length                     445277.0       5.993330       3.627044     0.00      3.000000      6.000000      10.00000  1.000000e+01
annual_inc                     466281.0   73277.381470   54963.568654  1896.00  45000.000000  63000.000000   88960.00000  7.500000e+06
dti                            466285.0      17.218758       7.851121     0.00     11.360000     16.870000      22.78000  3.999000e+01
delinq_2yrs                    466256.0       0.284678       0.797365     0.00      0.000000      0.000000       0.00000  2.900000e+01
inq_last_6mths                 466256.0       0.804745       1.091598     0.00      0.000000      0.000000       1.00000  3.300000e+01
mths_since_last_delinq         215934.0      34.104430      21.778487     0.00     16.000000     31.000000      49.00000  1.880000e+02
open_acc                       466256.0      11.187069       4.987526     0.00      8.000000     10.000000      14.00000  8.400000e+01
pub_rec                        466256.0       0.160564       0.510863     0.00      0.000000      0.000000       0.00000  6.300000e+01
revol_bal                      466285.0   16230.203487   20676.245152     0.00   6413.000000  11764.000000   20333.00000  2.568995e+06
revol_util                     465945.0      56.176947      23.732628     0.00     39.200000     57.600000      74.70000  8.923000e+02
total_acc                      466256.0      25.064430      11.600141     1.00     17.000000     23.000000      32.00000  1.560000e+02
out_prncp                      466285.0    4410.062342    6355.078769     0.00      0.000000    441.470000    7341.65000  3.216038e+04
out_prncp_inv                  466285.0    4408.452258    6353.198001     0.00      0.000000    441.380000    7338.39000  3.216038e+04
total_pymnt                    466285.0   11540.686220    8265.627112     0.00   5552.125349   9419.250943   15308.15846  5.777758e+04
total_pymnt_inv                466285.0   11469.892747    8254.157579     0.00   5499.250000   9355.430000   15231.31000  5.777758e+04
total_rec_prncp                466285.0    8866.014657    7031.687997     0.00   3708.560000   6817.760000   12000.00000  3.500003e+04
total_rec_int                  466285.0    2588.677225    2483.809661     0.00    957.280000   1818.880000    3304.53000  2.420562e+04
total_rec_late_fee             466285.0       0.650129       5.265730     0.00      0.000000      0.000000       0.00000  3.586800e+02
recoveries                     466285.0      85.344211     552.216084     0.00      0.000000      0.000000       0.00000  3.352027e+04
collection_recovery_fee        466285.0       8.961534      85.491437     0.00      0.000000      0.000000       0.00000  7.002190e+03
last_pymnt_amnt                466285.0    3123.913796    5554.737393     0.00    312.620000    545.960000    3187.51000  3.623444e+04
collections_12_mths_ex_med     466140.0       0.009085       0.108648     0.00      0.000000      0.000000       0.00000  2.000000e+01
policy_code                    466285.0       1.000000       0.000000     1.00      1.000000      1.000000       1.0000

- The average loan amount is around 14,300 dollars. However, this value may vary a lot, with a standard deviation of 8,286 dollars. The minimum is 500 dollars while the maximum is 35,000 dollars. Moreover, analyzing the IQR, 50% of the loan amounts are between 8,000 and 20,000 dollars. For funded amount and funded amount by investors we have quite similar values.
- The average interest rate is around 14%. The minimum is 5 while the maximum is 26. Analyzing the IQR, 50% of the interest rates are between 11% and 16.5%.
- . . .

- At a glance, there are some **variables** that can drive **valuable insights** about **credit risk** in Lending Club's operations:
    - Most of the demographical variables: addr_state, annual_inc, emp_length, home_ownership.
    - From the financial variables: dti, funded_amnt, installment, int_rate, loan_amnt, revol_bal, total_acc.
    - From the credit and risk variables: delinq_2yrs, fico scores variables, inq_last_6_months and 12m, and term.
    - From time-related and other variables: issue_d, purpose, tot_cur_bal, total_rev_hi_lim.